api/net/Areas/Subscriber/Controllers/ReportController.cs

```diff
@@ -0,0 +1,225 @@
+    {
+        var username = User.GetUsername() ?? throw new NotAuthorizedException("Username is missing");
+        var user = _userService.FindByUsername(username) ?? throw new NotAuthorizedException("User does not exist");
+        var result = _reportService.FindById(model.Id) ?? throw new InvalidOperationException("Report does not exist");
+        if (result?.OwnerId != user?.Id) throw new NotAuthorizedException("Not authorized to update this report");
+        _reportService.ClearChangeTracker(); // Remove the report from context.
+
+        result = _reportService.UpdateAndSave(model.ToEntity(_serializerOptions));
+        var report = _reportService.FindById(result.Id) ?? throw new InvalidOperationException("Report does not exist");
+        return new JsonResult(new ReportModel(report, _serializerOptions));
+    }
+
+    /// <summary>
+    /// Delete report for the specified 'id'.
+    /// </summary>
+    /// <param name="model"></param>
+    /// <returns></returns>
+    [HttpDelete("{id}")]
+    [Produces(MediaTypeNames.Application.Json)]
+    [ProducesResponseType(typeof(ReportModel), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(typeof(ErrorResponseModel), (int)HttpStatusCode.BadRequest)]
+    [SwaggerOperation(Tags = new[] { "Report" })]
+    public IActionResult Delete(ReportModel model)
+    {
+        var username = User.GetUsername() ?? throw new NotAuthorizedException("Username is missing");
+        var user = _userService.FindByUsername(username) ?? throw new NotAuthorizedException("User does not exist");
+        var result = _reportService.FindById(model.Id) ?? throw new InvalidOperationException("Report does not exist");
+        if (result?.OwnerId != user?.Id) throw new NotAuthorizedException("Not authorized to delete this report");
+        _reportService.ClearChangeTracker(); // Remove the report from context.
+
+        _reportService.DeleteAndSave(model.ToEntity(_serializerOptions));
+        return new JsonResult(model);
+    }
+
+    /// <summary>
+    /// Execute the report template and generate the results for previewing.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <returns></returns>
+    [HttpPost("{id}/preview")]
+    [Produces(MediaTypeNames.Application.Json)]
+    [ProducesResponseType(typeof(ReportResultModel), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(typeof(ErrorResponseModel), (int)HttpStatusCode.BadRequest)]
+    [SwaggerOperation(Tags = new[] { "Report" })]
+    public async Task<IActionResult> Preview(int id)
+    {
+        var username = User.GetUsername() ?? throw new NotAuthorizedException("Username is missing");
+        var user = _userService.FindByUsername(username) ?? throw new NotAuthorizedException("User does not exist");
+        var report = _reportService.FindById(id) ?? throw new InvalidOperationException("Report does not exist");
+        if (!user.Roles.Split(',').Contains(ClientRole.Administrator.GetName()) &&
```

Comment:
A subscriber can only preview their own reports.  Don't bother doing funky logic for admins.